hack/preload-images/preload_images.go

```diff
@@ -40,61 +42,106 @@ var (
 	dockerStorageDriver = "overlay2"
 	podmanStorageDriver = "overlay"
 	containerRuntimes   = []string{"docker", "containerd", "cri-o"}
-	k8sVersion          string
 	k8sVersions         []string
+	k8sVersion          = flag.String("kubernetes-version", "", "desired Kubernetes version, for example `v1.17.2`")
+	noUpload            = flag.Bool("no-upload", false, "Do not upload tarballs to GCS")
+	force               = flag.Bool("force", false, "Generate the preload tarball even if it's already exists")
+	limit               = flag.Int("limit", 0, "Limit the number of tarballs to generate")
 )
 
-func init() {
-	flag.StringVar(&k8sVersion, "kubernetes-version", "", "desired Kubernetes version, for example `v1.17.2`")
-	flag.Parse()
-	if k8sVersion != "" {
-		k8sVersions = append(k8sVersions, k8sVersion)
-	}
-	viper.Set("preload", "true")
+type preloadCfg struct {
+	k8sVer  string
+	runtime string
+}
+
+func (p preloadCfg) String() string {
+	return fmt.Sprintf("%q/%q", p.runtime, p.k8sVer)
 }
 
 func main() {
-	defer func() {
-		if err := deleteMinikube(); err != nil {
-			fmt.Printf("error cleaning up minikube: %v \n", err)
-		}
-	}()
+	flag.Parse()
+	viper.Set("preload", "true")
+
+	if *k8sVersion != "" {
+		k8sVersions = []string{*k8sVersion}
+	}
 
 	if err := deleteMinikube(); err != nil {
 		fmt.Printf("error cleaning up minikube at start up: %v \n", err)
 	}
 
-	if k8sVersions == nil {
-		var err error
-		k8sVersions, err = RecentK8sVersions()
-		if err != nil {
-			exit("Unable to get recent k8s versions: %v\n", err)
-		}
+	k8sVersions, err := collectK8sVers()
+	if err != nil {
+		exit("Unable to get recent k8s versions: %v\n", err)
 	}
 
-	k8sVersions = append(k8sVersions, constants.DefaultKubernetesVersion, constants.NewestKubernetesVersion, constants.OldestKubernetesVersion)
+	var toGenerate []preloadCfg
+	var i int
 
+out:
 	for _, kv := range k8sVersions {
 		for _, cr := range containerRuntimes {
-			tf := download.TarballName(kv, cr)
-			if download.PreloadExists(kv, cr) {
-				fmt.Printf("A preloaded tarball for k8s version %s - runtime %q already exists, skipping generation.\n", kv, cr)
-				continue
-			}
-			fmt.Printf("A preloaded tarball for k8s version %s - runtime %q doesn't exist, generating now...\n", kv, cr)
-			if err := generateTarball(kv, cr, tf); err != nil {
-				exit(fmt.Sprintf("generating tarball for k8s version %s with %s", kv, cr), err)
+			if *limit > 0 && i >= *limit {
+				break out
 			}
-			if err := uploadTarball(tf); err != nil {
-				exit(fmt.Sprintf("uploading tarball for k8s version %s with %s", kv, cr), err)
+			if *force || !download.PreloadExists(kv, cr) {
```

Comment:
add a comment what this will do.

Reply:
and also in the UI tell user, we are gonna overwrite...so we can see in the jenkins logs we are overwrtting